test/integration/revocation_test.go

```diff
@@ -33,6 +40,120 @@ func isPrecert(cert *x509.Certificate) bool {
+	if err != nil {
+		t.Fatalf("reading CA config: %s", err)
+	}
+
+	var conf struct {
+		CA struct {
+			Issuance struct {
+				Issuers []struct {
+					CRLURLBase string
+					CRLShards  int
+					Location   struct {
+						CertFile string
+					}
+				}
+			}
+		}
+	}
+
+	err = json.Unmarshal(b, &conf)
+	if err != nil {
+		t.Fatalf("unmarshaling CA config: %s", err)
+	}
+
+	ret := make(map[string][]*x509.RevocationList)
+
+	for _, issuer := range conf.CA.Issuance.Issuers {
+		issuerPEMBytes, err := os.ReadFile(issuer.Location.CertFile)
+		if err != nil {
+			t.Fatalf("reading CRL issuer: %s", err)
+		}
+
+		block, _ := pem.Decode(issuerPEMBytes)
+		issuerCert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			t.Fatalf("parsing CRL issuer: %s", err)
+		}
+
+		issuerSKID := hex.EncodeToString(issuerCert.SubjectKeyId)
+
+		// 10 is the number of shards configured in test/config*/crl-updater.json
+		for i := range 10 {
+			crlURL := fmt.Sprintf("%s%d.crl", issuer.CRLURLBase, i+1)
+			resp, err := http.Get(crlURL)
+			if err != nil {
+				t.Fatalf("getting CRL from %s: %s", crlURL, err)
+			}
+			// When there's nothing in a shard, I guess it doesn't get created?
+			if resp.StatusCode == 404 {
+				continue
+			}
+			body, err := io.ReadAll(resp.Body)
+			if err != nil {
+				t.Fatalf("reading CRL from %s: %s", crlURL, err)
+			}
+
+			list, err := x509.ParseRevocationList(body)
+			if err != nil {
+				t.Fatalf("parsing CRL from %s: %s (bytes: %x)", crlURL, err, string(body))
+			}
+
+			err = list.CheckSignatureFrom(issuerCert)
+			if err != nil {
+				t.Errorf("checking CRL signature on %s from %s: %s",
+					crlURL, issuerCert.Subject, err)
+			}
+
+			idpURIs, err := idp.GetIDPURIs(list.Extensions)
+			if err != nil {
+				t.Fatalf("getting IDP URIs: %s", err)
+			}
+			if len(idpURIs) != 1 {
+				t.Errorf("CRL at %s: expected 1 IDP URI, got %s", crlURL, idpURIs)
+			}
+			if idpURIs[0] != crlURL {
+				t.Errorf("fetched CRL from %s, got IDP of %s (should be same)", crlURL, idpURIs[0])
+			}
+
+			ret[issuerSKID] = append(ret[issuerSKID], list)
+		}
+	}
+	return ret
+}
+
+func checkRevoked(t *testing.T, revocations map[string][]*x509.RevocationList, cert *x509.Certificate, reason int) {
+	akid := hex.EncodeToString(cert.AuthorityKeyId)
+	if len(revocations[akid]) == 0 {
+		t.Errorf("no CRLs found for authorityKeyID %s", akid)
+	}
+	var matches []x509.RevocationListEntry
+	var count int
+	for _, list := range revocations[akid] {
+		for _, entry := range list.RevokedCertificateEntries {
+			count++
+			if entry.SerialNumber.Cmp(cert.SerialNumber) == 0 {
+				matches = append(matches, entry)
+			}
+		}
+	}
+	if len(matches) == 0 {
+		t.Errorf("didn't find matching entry on combined CRLs of length %d", count)
+
+	}
+	for _, match := range matches {
+		if match.ReasonCode != reason {
+			t.Errorf("revoked certificate %x: got reason %d, want %d", cert.SerialNumber, match.ReasonCode, reason)
+		}
+		return
+	}
```

Comment:
This return feels awkwardly placed. We're looping over and validating matches but returning on the first one, which doesn't seem like what we want to do.

Reply:
Another good catch. The intent was indeed to catch each possible mismatch; I think the misplaced `return` was an artifact of debugging.